difftool/diff/diff_tool.go

```diff
@@ -0,0 +1,220 @@
+	for _, a := range coreOrders {
+		keys = append(keys, a.Id)
+		coreStr = append(coreStr, a.String())
+	}
+
+	return diff("order", coreStr, dbIdToOrder, keys)
+}
+
+func diffDelegations(coreSnapshot *core.Snapshot, database *datanode.Database) error {
+	coreDelegations := coreSnapshot.GetDelegations()
+	dbDelegations, err := database.GetDelegations()
+	if err != nil {
+		return err
+	}
+
+	dbIdToDelegation := make(map[string]string, len(dbDelegations))
+	for _, ab := range dbDelegations {
+		dbProto := ab.ToProto()
+		dbIdToDelegation[dbProto.Party+dbProto.NodeId+dbProto.EpochSeq] = dbProto.String()
+	}
+
+	keys := make([]string, 0, len(coreDelegations))
+	coreStr := make([]string, 0, len(coreDelegations))
+	for _, a := range coreDelegations {
+		keys = append(keys, a.Party+a.NodeId+a.EpochSeq)
+		coreStr = append(coreStr, a.String())
+	}
+
+	return diff("delegation", coreStr, dbIdToDelegation, keys)
+}
+
+func diffProposals(coreSnapshot *core.Snapshot, database *datanode.Database) error {
+	coreProposals := coreSnapshot.GetProposals()
+	dbProposals, err := database.GetProposals()
+	if err != nil {
+		return err
+	}
+
+	dbIdToProposal := make(map[string]string, len(dbProposals))
+	for _, ab := range dbProposals {
+		dbProto := ab.ToProto()
+		dbIdToProposal[dbProto.Id] = dbProto.String()
+	}
+
+	keys := make([]string, 0, len(coreProposals))
+	coreStr := make([]string, 0, len(coreProposals))
+	for _, a := range coreProposals {
+		keys = append(keys, a.Id)
+		coreStr = append(coreStr, a.String())
+	}
+
+	return diff("proposal", coreStr, dbIdToProposal, keys)
+}
+
+func diffMarkets(coreSnapshot *core.Snapshot, database *datanode.Database) error {
+	coreMarkets := coreSnapshot.GetMarkets()
+	dbMarkets, err := database.GetMarkets()
+	if err != nil {
+		return err
+	}
+
+	dbIdToMarket := make(map[string]string, len(dbMarkets))
+	for _, ab := range dbMarkets {
+		dbProto := ab.ToProto()
+		dbIdToMarket[dbProto.Id] = dbProto.String()
+	}
+
+	keys := make([]string, 0, len(coreMarkets))
+	coreStr := make([]string, 0, len(coreMarkets))
+	for _, a := range coreMarkets {
+		keys = append(keys, a.Id)
+		coreStr = append(coreStr, a.String())
+	}
+
+	return diff("makret", coreStr, dbIdToMarket, keys)
+}
+
+func diffTransfers(coreSnapshot *core.Snapshot, database *datanode.Database) error {
+	coreTransfers := coreSnapshot.GetTransfers()
+	dbTransfers, err := database.GetTransfers()
+
+	if err != nil {
+		return err
+	}
+
+	dbIdToTransfer := make(map[string]string, len(dbTransfers))
+	for _, ab := range dbTransfers {
+		dbIdToTransfer[ab.Id] = ab.String()
+	}
+
+	keys := make([]string, 0, len(coreTransfers))
+	coreStr := make([]string, 0, len(coreTransfers))
+	for _, a := range coreTransfers {
+		keys = append(keys, a.Id)
+		coreStr = append(coreStr, a.String())
+	}
+
+	return diff("proposal", coreStr, dbIdToTransfer, keys)
+}
+
+func diff(name string, core []string, db map[string]string, key []string) error {
+	errs := []string{}
+	for i, coreStr := range core {
```

Comment:
Aren't we going to miss cases where something is in the data-node, but missing from the core-snapshot? i.e the core snapshot could be completely empty and the tool will say its fine, there is no diff.

Reply:
1. don't waste time looking at *this* version. let me push the new one. 
2. the purpose here to check the sanity of datanode treating the snapshot as the ground truth

Reply:
I just pushed the latest version